### PR TITLE
Multiplayer input support for the Wii

### PR DESCRIPTION
**Pull request by camthehaxman**
This small change allows co-op mode to work. If playing as Sonic and Tails, a second Wii remote can be connected and will control Tails.

I still haven't yet figured out how to get competition mode working. Even after connecting multiple Wii remotes the character select screen only responds to the first one.

### CODE REVIEW

**Comment by camthehaxman** [comment]
I've gotten the input to work properly. However, split screen doesn't work in competition mode. It only shows Player 1's screen.

**Comment by Mefiresu** [comment]
> I've gotten the input to work properly. However, split screen doesn't work in competition mode. It only shows Player 1's screen.

The current code only draws the main (first) framebuffer to the screen.
The GX code would require some modifications in order to be able to copy and draw up to 4 framebuffers on the screen (with a GX equivalent of https://github.com/Mefiresu/RSDKv5-Decompilation/blob/7b21415279f0abdbb356688581b46061c2b507a0/RSDKv5/RSDK/Graphics/GLFW/GLFWRenderDevice.cpp#L343), along with the removal of the SCREEN_COUNT macro redefinition that was done to save RAM https://github.com/Mefiresu/RSDKv5-Decompilation/blob/7b21415279f0abdbb356688581b46061c2b507a0/platforms/NintendoWii.cmake#L50.
I have no idea if there's enough RAM left to fit split screen multiplayer.

**Comment by camthehaxman** [comment]
I'm kind of confused as to how the drawing actually works. As I understand it, everything is software rendered to a 424x240 framebuffer, and the GX backend converts that framebuffer to a texture and draws it as a quad onto the screen. There's some weird tiling stuff going on [here](https://github.com/Mefiresu/RSDKv5-Decompilation/blob/dev/wii-port/RSDKv5/RSDK/Graphics/GX/GXRenderDevice.cpp#L323), and I'm not entirely sure why that's necessary. The Wii natively supports RGB565 textures.

**Comment by Mefiresu** [comment]
The tiling is done to improve widescreen resolution (instead of squishing the texture to 640x480 for anamorphic widescreen, multiple draw calls are used to use the entire 720x480 framebuffer).

**Comment by camthehaxman** [comment]
Ok, I'll probably investigate split screen in a separate PR, then. This one at least allows co-op to work, which didn't before.

**Comment by camthehaxman** [comment]
@Mefiresu Any other concerns with this PR, or can it be merged?

File: RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp [hunk 1]
```diff
@@ -129,12 +131,15 @@ void RSDK::SKU::InputDeviceWii::ProcessInput(int32 controllerID) {
 }
 
 void RSDK::SKU::InputDeviceWii::CloseDevice() {
+    //printf("CloseDevice(): closing wii input device %i\r");
     this->active     = false;
     this->isAssigned = false;
-    WPAD_Shutdown();
```
Comment (Mefiresu) [inline RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp:134]: Could this be readded to `RenderDevice::ProcessEvents()` maybe? (after setting `isRunning = false`, which is essentially triggering an exit)

File: RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp [hunk 2]
```diff
@@ -129,12 +131,15 @@ void RSDK::SKU::InputDeviceWii::ProcessInput(int32 controllerID) {
 }
 
 void RSDK::SKU::InputDeviceWii::CloseDevice() {
+    //printf("CloseDevice(): closing wii input device %i\r");
```
Comment (Mefiresu) [inline RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp:134]: To be removed?

File: RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp [hunk 3]
```diff
@@ -129,12 +131,15 @@ void RSDK::SKU::InputDeviceWii::ProcessInput(int32 controllerID) {
...
     this->active     = false;
     this->isAssigned = false;
-    WPAD_Shutdown();
 }
 
 RSDK::SKU::InputDeviceWii *RSDK::SKU::InitWiiInputDevice(uint32 id) {
+    /*printf("InitWiiInputDevice(%u): new %s on port %i\r",
+        id, SKU::IsGameCubeController(id) ? "GameCube controller" : "wiimote", SKU::GetWiiInputDevicePort(id));*/
+
```
Comment (Mefiresu) [inline RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp:142]: To be removed?

File: RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp [hunk 4]
```diff
@@ -165,7 +170,7 @@ RSDK::SKU::InputDeviceWii *RSDK::SKU::InitWiiInputDevice(uint32 id) {
 }
 
 void RSDK::SKU::InitWiiInputAPI() {
+    printf("InitWiiInputAPI()\r");
```
Comment (Mefiresu) [inline RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp:173]: To be removed?

**Review by Mefiresu — COMMENTED** [review]
This looks fine, I'll try to test it on my end soon and will either merge or request other changes if needed.

File: RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp (hunk 1 above)
Comment (camthehaxman) [reply RSDKv5/RSDK/Input/Wii/WiiInputDevice.cpp:134]: I did what the SDL2 input backend does and added a `ReleaseWiiInputAPI` function that gets called when the engine exits.

**Comment by Mefiresu** [comment]
Works fine on my end, thank you!

**Pull request merged**